src/core/DataFile.cpp

```diff
@@ -1386,6 +1387,83 @@ void DataFile::upgrade_noHiddenClipNames()
+		for (int i = 0; !tracks.item(i).isNull(); i++)
+		{
+			// Ignore BB container tracks
+			if (tracks.item(i).toElement().attribute("type").toInt() == 1) { continue; }
+
+			QDomNodeList instruments = tracks.item(i).toElement().elementsByTagName("instrument");
+			if (instruments.isEmpty()) { continue; }
+			QDomElement instrument = instruments.item(0).toElement();
+			// Raise the base note of every instrument by 12 to compensate for the change
+			// of A4 key code from 57 to 69. This ensures that notes are labeled correctly.
+			instrument.parentNode().toElement().setAttribute(
+				"basenote",
+				instrument.parentNode().toElement().attribute("basenote").toInt() + 12);
+			// Raise the pitch of all notes in patterns assigned to instruments not affected
+			// by #1857 by an octave. This negates the base note change for normal instruments,
+			// but leaves the MIDI-based instruments sounding an octave lower, preserving their
+			// pitch in existing projects.
+			if (instrument.attribute("name") != "zynaddsubfx" &&
+				instrument.attribute("name") != "vestige" &&
+				instrument.attribute("name") != "lv2instrument" &&
+				instrument.attribute("name") != "carlapatchbay" &&
+				instrument.attribute("name") != "carlarack")
+			{
+				QDomNodeList patterns = tracks.item(i).toElement().elementsByTagName("pattern");
+				for (int i = 0; !patterns.item(i).isNull(); i++)
+				{
+					QDomNodeList notes = patterns.item(i).toElement().elementsByTagName("note");
+					for (int i = 0; !notes.item(i).isNull(); i++)
+					{
+						notes.item(i).toElement().setAttribute(
+							"key",
+							notes.item(i).toElement().attribute("key").toInt() + 12
+						);
+					}
+				}
+			}
+		}
+	}
+	else
+	{
+		// Dealing with a preset, not a song
+		QDomNodeList presets = elementsByTagName("instrumenttrack");
+		if (presets.item(0).isNull()) { return; }
+		QDomElement preset = presets.item(0).toElement();
+		// Common correction for all instrument presets (make base notes match the new MIDI range).
+		// NOTE: Many older presets do not have any basenote defined, assume they were "made for 57".
+		// (Specifying a default like this also happens to fix a FileBrowser bug where previews of presets
+		// with undefined basenote always play with the basenote inherited from previously played preview.)
+		int oldBase = preset.attribute("basenote", "57").toInt();
+		preset.setAttribute("basenote", oldBase + 12);
+		// Extra correction for Zyn, VeSTige, LV2 and Carla (to preserve the original buggy behavior).
+		QDomNodeList instruments = presets.item(0).toElement().elementsByTagName("instrument");
+		if (instruments.isEmpty()) { return; }
+		QDomElement instrument = instruments.item(0).toElement();
+		if (instrument.attribute("name") == "zynaddsubfx" ||
+			instrument.attribute("name") == "vestige" ||
+			instrument.attribute("name") == "lv2instrument" ||
+			instrument.attribute("name") == "carlapatchbay" ||
+			instrument.attribute("name") == "carlarack")
```

Comment:
This check is long and repeated twice, so it should be simplified or split out if possible.

Reply:
Good idea, fixed. (Although, ironically, by eliminating the repetition the number of lines actually _increased_ by one due to the overhead of our coding conventions.. :smile: )